### PR TITLE
Feature/wstat

### PR DESCRIPTION
**Pull request by dtnguyen2**
# Release Notes
Implement wstat as described at the following url:
https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSappendixStatistics.html

# Description
The bkg file associated with the response file must be available in order to use wstat,
an exception is raised if bkg file is not available.
This PR is based on #90

### CODE REVIEW

File: sherpa/stats/__init__.py [hunk 1]
```diff
@@ -474,29 +492,93 @@ def __getstate__(self):
...
+    .. [1] The description of the W statistic (`wstat`) in
+           https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSappendixStatistics.html
+
+    """
+
+    def __init__(self, name='wstat'):
+        Likelihood.__init__(self, name)
+
+    @staticmethod
+    def calc_stat(data, model, staterror=None, syserror=None,
```
Comment (cdeil) [inline sherpa/stats/__init__.py]: Thanks for implementing this!

As a future user of this feature, I'd like to ask / comment:

This is shoehorning wstat into a `Likelihood.calc_stat` interface that was designed for chi2 stats, right?
I.e. arguments are re-used for something completely different than their name says: `exposuretime = syserror`

I guess there is no better way to get wstat into the existing API?
If so, can you at least explain what's going on here in the docstring?

Comment (dtnguyen2) [inline sherpa/stats/__init__.py]: If I make the following changes:

    @staticmethod
    def calc_stat(data, model, datasize=None, exposuretime=None,
                  weight=None, **kwargs):

        if 'bkg' in kwargs.keys():
            bkg = kwargs.get('bkg')
        else:
            raise FitErr('no bkg file is supplied, use cstat instead')
        return _statfcts.calc_wstat_stat(data, model, datasize,
exposuretime,
                                         bkg, truncation_value)

Then it breaks symmetry with the other classes within the stats/__init__.py
file and we currently call calc_stat with 4 arguments (without keywords)
but I don't think I'm totally against this.  Hopefully no future
modifications will require the staterror and syserror keywords.

BTW, I thought about adding  two additional arguments datasize and
exposuretime to the calc_stat method of the various Statistic classes but
the problem with this approach is that it breaks backwards compatibility
with the user's existing user-stat scripts or it  makes the logical if
statement in fit.py too complicated and be error prone.

File: sherpa/fit.py [hunk 2]
```diff
@@ -470,9 +471,36 @@ def __init__(self, data, model, stat, method, itermethod_opts={'name': 'none'}):
...
+            mydata = self.data.datasets[index]
+            if hasattr(mydata, 'response_ids') and \
+                    hasattr(mydata, 'background_ids') and \
+                    len(mydata.response_ids) and len(mydata.background_ids):
+                bkg = mydata.get_background(mydata.background_ids[0])
+                tmp_bkg_dep = bkg.get_dep(True)
+                data_size[index] = tmp_bkg_dep.size
+                bkg_dep = append(bkg_dep, tmp_bkg_dep)
+                exposure_time[2 * index] = mydata.exposure
+                exposure_time[2 * index + 1] = bkg.exposure
```
Comment (DougBurke) [inline sherpa/fit.py]: For the wstat case, a simple hack to include the BACKSCAL ratios, would be

```
                try:
                    bexp = bkg.exposure * bkg.backscal / mydata.backscal
                except AttributeError:
                    bexp = bkg.exposure
                exposure_time[2 * index + 1] = bexp
```

which seems to give me the same stat values as XSPEC does. I do not like this solution, as (as also mentioned by @cdell) this feels like trying to work around the limitations of an interface, but I don't have any concrete suggestions for improvements.

**Comment by DougBurke** [comment]
As mentioned by @anetasie we need to add WStat support to the MCMC/pyBLoCXS code. I had to add `WStat` to https://github.com/sherpa/sherpa/blob/f4e8aa346a9f577a8d435eb46a6fe85015dc9e18/sherpa/sim/__init__.py#L27 and https://github.com/sherpa/sherpa/blob/f4e8aa346a9f577a8d435eb46a6fe85015dc9e18/sherpa/sim/__init__.py#L494 (and the text of the error message on the next line)

**Comment by DougBurke** [comment]
And added to https://github.com/sherpa/sherpa/blob/f4e8aa346a9f577a8d435eb46a6fe85015dc9e18/sherpa/plot/__init__.py#L76

**Comment by DougBurke** [comment]
There is a problem when re-grouping the data (I have not checked against yesterday's check in yet). If the grouping is changed then you get an error saying

    TypeError: statistic input array sizes do not match

whenever the statistic is next calculated (e.g. by `fit` or `calc_stat`). An example script is shown below. My guess is that the grouping isn't being applied to the background (and, if so,  it may be tricky to work out how best to handle this).

# Example

The program

```
% cat doit.py
from sherpa.astro import ui
ui.load_pha(1, 'sherpa/astro/datastack/tests/data/3c273.pi')
ui.set_stat('wstat')
ui.set_method('simplex')
ui.notice_id(1, 0.5, 7)
ui.set_source(1, ui.xsphabs.gal * ui.xspowerlaw.pl)
pl.norm = 1e-4
s1 = ui.calc_stat(1)
print("stat 1 = {}".format(s1))
ui.group_counts(1, 5)
s2 = ui.calc_stat(1)
print("stat 2 = {}".format(s2))
```

gives

```
% python doit.py
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find xpaget on your PATH'
WARNING: systematic errors were not found in file 'sherpa/astro/datastack/tests/data/3c273.pi'
statistical errors were found in file 'sherpa/astro/datastack/tests/data/3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file sherpa/astro/datastack/tests/data/3c273.arf
read RMF file sherpa/astro/datastack/tests/data/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa/astro/datastack/tests/data/3c273_bg.pi'
statistical errors were found in file 'sherpa/astro/datastack/tests/data/3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file sherpa/astro/datastack/tests/data/3c273_bg.pi
stat 1 = 1081.43303654
Traceback (most recent call last):
  File "doit.py", line 11, in <module>
    s2 = ui.calc_stat(1)
  File "<string>", line 1, in calc_stat
  File "/lagado.real/sherpa/sherpa-wstat/sherpa/ui/utils.py", line 7830, in calc_stat
    return f.calc_stat()
  File "/lagado.real/sherpa/sherpa-wstat/sherpa/fit.py", line 990, in calc_stat
    bkg=bkg_data)[0]
  File "/lagado.real/sherpa/sherpa-wstat/sherpa/stats/__init__.py", line 585, in calc_stat
    truncation_value)
TypeError: statistic input array sizes do not match
```

If I manually hack the background to have the same grouping as the data then it works. Adding in the following after the `group_counts` call means that the program runs:

````
d = ui.get_data(1)
b = ui.get_bkg(1) 
b.grouping = d.grouping
b.mask = d.mask
```

which gives

```
% python doit2.py
...
stat 1 = 1081.43303654
stat 2 = 1161.32718342
```

**Comment by dtnguyen2** [comment]
I'm pretty sure this behavior has been around for a while. When you sent
out the bug report about wstat misbehaving after calc_data_sum, I had a
fixed for it forcing all filter and masks for data and background to match
but we decided that we will address this issue outside of wstat etc... So
perhaps we should think about fixing group instead?

**Comment by dtnguyen2** [comment]
I implemented the wstat verbatim from the following url:
heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSappendixStatistics.html
However, recently I stumbled on the following url from K. Arnaud:
heasarc.gsfc.nasa.gov/docs/xanadu/xspec/wstat.ps
On page 5, it says something to the effect: 
"...to give a quadratic in b_i whose solution is:....."
I have to look at this more closely but it looks like he is using the grade school formula 
(-b +/- sqrt(b^2-4ac))/2a for one of the roots (see equation a_i) of quadratic equation which is not numerically stable.

**Comment by DougBurke** [comment]
@dtnguyen2 one interesting project on converting equations into more-stable forms is herbie http://herbie.uwplse.org/ (here's one link to it http://cs.stackexchange.com/questions/44263/program-transformations-for-numeric-stability but I've been tracking it for some time, just never got around to doing anything with it).

**Pull request merged**